### PR TITLE
Refactor the structure of data transform design.

### PR DESCRIPTION
**Pull request by brightcoder01**
Complement the overview for data transform design. Add Background/Challenge/Related Work sections.
The root issue of the ongoing discussion series is #1670

### CODE REVIEW

File: docs/designs/data_transform.md [hunk 1]
```diff
@@ -1,16 +1,38 @@
...
+[SQLFlow](https://github.com/sql-machine-learning/sqlflow) is a bridge that connects a SQL engine and machine learning toolkits. It extends the SQL syntax to define a machine learning pipeline. Naturally SQLFlow should be able to describe the data transform process. In this doc, we are focusing on how to do data transform using SQLFlow.  
 
-Consistency between offline and online is the key point of data transform. Users write the transform code only once. And then the same logic can run in batch mode for training and in real time mode for serving. In this way, we can prevent the training/serving skew. Both TF Transform and feature column can keep the consistency. The data transform logic in the training stage is built into the inference graph as the SavedModel.  
+## Challenge
 
-## Transform Expression in SQLFlow
+* In some systems, user implement the transform code separately for training and serving, it may bring the training/serving skew. Consistency between training and serving is the key point of data transform. Users write the transform code only once. And then the same logic can run in batch mode for training and in real time mode for serving.
```
Comment (workingloong) [inline docs/designs/data_transform.md]: "user implement -> the user implements " or "user -> users"

File: docs/designs/data_transform.md [hunk 2]
```diff
@@ -1,16 +1,38 @@
...
 
-Consistency between offline and online is the key point of data transform. Users write the transform code only once. And then the same logic can run in batch mode for training and in real time mode for serving. In this way, we can prevent the training/serving skew. Both TF Transform and feature column can keep the consistency. The data transform logic in the training stage is built into the inference graph as the SavedModel.  
+## Challenge
 
-## Transform Expression in SQLFlow
+* In some systems, user implement the transform code separately for training and serving, it may bring the training/serving skew. Consistency between training and serving is the key point of data transform. Users write the transform code only once. And then the same logic can run in batch mode for training and in real time mode for serving.  
+* The transform logic is incomplete at the time when user develops the pipeline. The statistical value on the entire dataset at runtime is necessary to make the transform logic concrete.
```
Comment (workingloong) [inline docs/designs/data_transform.md]: Remove "at the time"?

File: docs/designs/data_transform.md [hunk 3]
```diff
@@ -1,16 +1,38 @@
...
+[SQLFlow](https://github.com/sql-machine-learning/sqlflow) is a bridge that connects a SQL engine and machine learning toolkits. It extends the SQL syntax to define a machine learning pipeline. Naturally SQLFlow should be able to describe the data transform process. In this doc, we are focusing on how to do data transform using SQLFlow.  
 
-Consistency between offline and online is the key point of data transform. Users write the transform code only once. And then the same logic can run in batch mode for training and in real time mode for serving. In this way, we can prevent the training/serving skew. Both TF Transform and feature column can keep the consistency. The data transform logic in the training stage is built into the inference graph as the SavedModel.  
+## Challenge
```
Comment (workingloong) [inline docs/designs/data_transform.md:10]: How to write transform code using SQL syntax maybe also a challenge?

File: docs/designs/data_transform.md [hunk 4]
```diff
@@ -1,16 +1,38 @@
...
+From users perspective, SQLFlow users prefer to write SQL instead of python. It's user unfriendly to SQLFlow users if we integrate TF Transform with SQLFlow directly.  
+
+### SQL
+
+User can write SQL to do the analysis and transform work with [built-in functions](https://www.alibabacloud.com/help/doc-detail/96342.htm?spm=a2c63.p38356.b99.111.27e27309rgC5m1) or [UDF(User Defined Function)](https://www.alibabacloud.com/blog/udf-development-guide-with-maxcompute-studio_594738). But we can't use SQL to transform the data for serving. We need implement the transform code again using C++ or other high performance language. It may bright the training/serving skew.
```
Comment (workingloong) [inline docs/designs/data_transform.md]: Why do we use c++？ In my opinion, we should use TensorFlow ops not a language.

File: docs/designs/data_transform.md [hunk 5]
```diff
@@ -1,16 +1,38 @@
...
+
+### Internal System
+
+The library in the internal system is configure driven. It contains some primitive transform ops and user compose the transform logic with configuration file. The inference engine is compiled with the internal library. User need both feature engineering configuration files and SavedModel to deploy the model for serving.  
+
+## Our Approach
+
+From the point of view from SQLFlow, SQL can naturally support statistical work just like the analyzer. [Feature column API](https://tensorflow.google.cn/api_docs/python/tf/feature_column) and [keras preprocessing layer](https://github.com/tensorflow/community/pull/188) can take charge of transform logic as transformer. For dense column, we can use numeric_column and pass a user defined function *normalizer_fn* to convert the column value. For sparse column, we can use embedding_column to map the sparse value to embedding vector or use cross_column to do the feature crossing. We plan to use SQL and feature column/keras preprocessing layer together to do the data transform work.
```
Comment (workingloong) [inline docs/designs/data_transform.md]: I was wondering whether there is sparse column in the table? Do the sparse column represent the column with ids string like ["1,2,4", "1,5,6,7"]. But it is also a dense column as string format. What's more, tf-serving does not support varied length tensor.

File: docs/designs/data_transform.md (hunk 1 above)
Comment (brightcoder01) [reply docs/designs/data_transform.md]: Good catch, updated.

File: docs/designs/data_transform.md (hunk 2 above)
Comment (brightcoder01) [reply docs/designs/data_transform.md]: Good catch. updated.

File: docs/designs/data_transform.md (hunk 3 above)
Comment (brightcoder01) [reply docs/designs/data_transform.md:10]: Good suggestion, added.

File: docs/designs/data_transform.md (hunk 4 above)
Comment (brightcoder01) [reply docs/designs/data_transform.md]: The main point is that SQL is just suitable for batch processing not for serving.
Rephrase it to be "use other programming language"

File: docs/designs/data_transform.md (hunk 5 above)
Comment (brightcoder01) [reply docs/designs/data_transform.md]: The sparse column represent the column feed into categorical_column_with_xxx or sequence_categorical_column_with_xxx. It can be tf.string or tf.int32 and so on.

File: docs/designs/data_transform.md [hunk 6]
```diff
@@ -1,19 +1,42 @@
...
 
-Consistency between offline and online is the key point of data transform. Users write the transform code only once. And then the same logic can run in batch mode for training and in real time mode for serving. In this way, we can prevent the training/serving skew. Both TF Transform and feature column can keep the consistency. The data transform logic in the training stage is built into the inference graph as the SavedModel.  
+## Challenge
 
-## Transform Expression in SQLFlow
+* In some systems, users implement the transform code separately for both training and serving, it may bring the training/serving skew. Consistency between training and serving is the key point of data transform. Users write the transform code only once. And then the same logic can run in batch mode for training and in real time mode for serving.  
+* The transform logic is incomplete when users develop the pipeline. The statistical value calculated from the entire dataset at runtime is necessary to make the transform logic concrete.  
+* Express the feature engineering operations with SQL extended statement.
```
Comment (QiJune) [inline docs/designs/data_transform.md]: `Express the feature engineering operations with SQL extended statement`
This is not a complete sentence.

Comment (brightcoder01) [reply docs/designs/data_transform.md]: Good catch, fixed.

File: docs/designs/data_transform.md [hunk 7]
```diff
@@ -1,30 +1,79 @@
...
+Data transform is an important part in an end-to-end machine learning pipeline. It processes the raw data using operations such as [standardization, normalization](https://en.wikipedia.org/wiki/Feature_scaling), [bucketization](https://en.wikipedia.org/wiki/Data_binning) and so on. The target is to make sure the data is in the right format and ready for both model training and inference. Data transform contains two key parts: analyzer and transformer. Analyzer scans the entire data set and calculates the statistical values such as mean, min, variance and so on. Transformer combines the statistical value and the transform function to construct the concrete transform logic. And then it transforms the data records one by one. The transform logic should be consistent between training and inference.
```
Comment (skydoorkai) [inline docs/designs/data_transform.md]: The target is to make sure -> After data transform, the data is ...

and so on -> , etc.

uses statistical values if any and transform functions to ...

File: docs/designs/data_transform.md [hunk 8]
```diff
@@ -1,30 +1,79 @@
...
+[SQLFlow](https://github.com/sql-machine-learning/sqlflow) is a bridge that connects a SQL engine and machine learning toolkits. It extends the SQL syntax to define a machine learning pipeline. Naturally SQLFlow should be able to describe the data transform process. In this doc, we are focusing on how to do data transform using SQLFlow.  
 
-Consistency between offline and online is the key point of data transform. Users write the transform code only once. And then the same logic can run in batch mode for training and in real time mode for serving. In this way, we can prevent the training/serving skew. Both TF Transform and feature column can keep the consistency. The data transform logic in the training stage is built into the inference graph as the SavedModel.  
+## Challenge
 
-## Transform Expression in SQLFlow
+* In some ML systems, users implement the transform code separately for both training and serving. It may bring the training/serving skew. Consistency between training and serving is the key point of data transform. Users write the transform code only once. And then the same logic can run in batch mode for training and in real time mode for serving.
```
Comment (skydoorkai) [inline docs/designs/data_transform.md]: users use different codes for training and serving. This may cause inconsistency between training and serving.

File: docs/designs/data_transform.md [hunk 9]
```diff
@@ -1,30 +1,79 @@
...
 
-Consistency between offline and online is the key point of data transform. Users write the transform code only once. And then the same logic can run in batch mode for training and in real time mode for serving. In this way, we can prevent the training/serving skew. Both TF Transform and feature column can keep the consistency. The data transform logic in the training stage is built into the inference graph as the SavedModel.  
+## Challenge
 
-## Transform Expression in SQLFlow
+* In some ML systems, users implement the transform code separately for both training and serving. It may bring the training/serving skew. Consistency between training and serving is the key point of data transform. Users write the transform code only once. And then the same logic can run in batch mode for training and in real time mode for serving.  
+* The transform logic is incomplete when users develop the pipeline. The statistical value calculated from the entire dataset at runtime is necessary to make the transform logic concrete.
```
Comment (skydoorkai) [inline docs/designs/data_transform.md]: The meaning of this sentence is not very clear to me.

File: docs/designs/data_transform.md [hunk 10]
```diff
@@ -1,30 +1,79 @@
...
+
+## Our Approach
+
+From the perspective of SQLFLow, SQL can naturally support statistical work just like the analyzer. [Feature column API](https://tensorflow.google.cn/api_docs/python/tf/feature_column) and [keras preprocessing layer](https://github.com/tensorflow/community/pull/188) can take charge of transform work as transformer. For dense column, we can use [numeric_column](https://www.tensorflow.org/api_docs/python/tf/feature_column/numeric_column) and pass a user defined function *normalizer_fn* to convert the column value. For sparse column, we can use [embedding_column](https://www.tensorflow.org/api_docs/python/tf/feature_column/embedding_column) to map the sparse value to embedding vector or use [cross_column](https://www.tensorflow.org/api_docs/python/tf/feature_column/crossed_column) to do the feature crossing. We plan to use SQL and feature column/keras preprocessing layer together to do the data transform work.
```
Comment (skydoorkai) [inline docs/designs/data_transform.md]: transform work -> the transform work

Comment (skydoorkai) [reply docs/designs/data_transform.md]: to embedding vector -> to an embedding vector

File: docs/designs/data_transform.md [hunk 11]
```diff
@@ -1,30 +1,79 @@
...
+
+## Our Approach
+
+From the perspective of SQLFLow, SQL can naturally support statistical work just like the analyzer. [Feature column API](https://tensorflow.google.cn/api_docs/python/tf/feature_column) and [keras preprocessing layer](https://github.com/tensorflow/community/pull/188) can take charge of transform work as transformer. For dense column, we can use [numeric_column](https://www.tensorflow.org/api_docs/python/tf/feature_column/numeric_column) and pass a user defined function *normalizer_fn* to convert the column value. For sparse column, we can use [embedding_column](https://www.tensorflow.org/api_docs/python/tf/feature_column/embedding_column) to map the sparse value to embedding vector or use [cross_column](https://www.tensorflow.org/api_docs/python/tf/feature_column/crossed_column) to do the feature crossing. We plan to use SQL and feature column/keras preprocessing layer together to do the data transform work.  
+For the consistency between training and serving, both feature column and keras preprocessing layer can make it. The data transform logic in the training stage is built into the inference graph using the SavedModel format.
```
Comment (skydoorkai) [inline docs/designs/data_transform.md]: Both feature column and Keras preprocessing layer can guarantee the consistency between training and serving.

File: docs/designs/data_transform.md [hunk 12]
```diff
@@ -1,30 +1,79 @@
 # Data Tranform Design Doc
 
-## Motivation
+## Backgroud
```
Comment (terrytangyuan) [inline docs/designs/data_transform.md]: Background

File: docs/designs/data_transform.md (hunk 7 above)
Comment (terrytangyuan) [inline docs/designs/data_transform.md]: Does this have to be the entire dataset? Are we planning to support streaming data? e.g. statistics like mean, min, var can also be obtained via a streaming fashion

File: docs/designs/data_transform.md (hunk 8 above)
Comment (terrytangyuan) [inline docs/designs/data_transform.md]: data transform -> data transformation
transform -> transformation

File: docs/designs/data_transform.md [hunk 13]
```diff
@@ -1,30 +1,79 @@
...
+## Related Work
+
+### TensorFlow Transform
+
+[TensorFlow Transform](https://www.tensorflow.org/tfx/transform/get_started) is the open source solution for data transform in [TensorFlow Extended](https://www.tensorflow.org/tfx/guide). Users need write a python function `preprocess_fn` to define the preprocess logic. The preprocessing function contains two group of API calls: TensorFlow Transform Analyzers and TensorFlow Ops. Analyzer will do the statistical work on the training dataset once and convert to constant tensors. And then the statistical value and TensorFlow Ops will make the concrete transform logic as a TensorFlow graph to convert the record one by one. The graph will be used for both training and serving.
```
Comment (terrytangyuan) [inline docs/designs/data_transform.md]: python -> Python

File: docs/designs/data_transform.md [hunk 14]
```diff
@@ -1,30 +1,79 @@
...
+## Related Work
+
+### TensorFlow Transform
+
+[TensorFlow Transform](https://www.tensorflow.org/tfx/transform/get_started) is the open source solution for data transform in [TensorFlow Extended](https://www.tensorflow.org/tfx/guide). Users need write a python function `preprocess_fn` to define the preprocess logic. The preprocessing function contains two group of API calls: TensorFlow Transform Analyzers and TensorFlow Ops. Analyzer will do the statistical work on the training dataset once and convert to constant tensors. And then the statistical value and TensorFlow Ops will make the concrete transform logic as a TensorFlow graph to convert the record one by one. The graph will be used for both training and serving.  
+Let's take [normalizing(min-max normalization)](https://en.wikipedia.org/wiki/Feature_scaling) the column value `capital_gain` in [census income dataset](https://archive.ics.uci.edu/ml/datasets/Census+Income) for example. The following is the `preprocess_fn` definition with TensorFlow Transform:
```
Comment (terrytangyuan) [inline docs/designs/data_transform.md]: missing space before "(min-max"

File: docs/designs/data_transform.md [hunk 15]
```diff
@@ -1,30 +1,79 @@
...
+```python
+import tensorflow_transform as tft
+
+def preprocess_fn(inputs):
+    outputs = inputs.copy()
+    outputs["capital_gain"] = tft.scale_to_0_1(inputs["capital_gain"])
+    return outputs
+```
+
+From users' perspective, SQLFlow users prefer to write SQL instead of python. It's user unfriendly if we integrate TF Transform with SQLFlow directly.
```
Comment (terrytangyuan) [inline docs/designs/data_transform.md]: user unfriendly -> not user-friendly

File: docs/designs/data_transform.md (hunk 7 above)
Comment (brightcoder01) [reply docs/designs/data_transform.md]: Good suggestion. Updated.
The target is to make sure -> After transformation, the data is...

File: docs/designs/data_transform.md (hunk 10 above)
Comment (brightcoder01) [reply docs/designs/data_transform.md]: Good suggestion, updated.

File: docs/designs/data_transform.md (hunk 11 above)
Comment (brightcoder01) [reply docs/designs/data_transform.md]: Good suggestion, updated.

File: docs/designs/data_transform.md (hunk 12 above)
Comment (brightcoder01) [reply docs/designs/data_transform.md]: Good catch, updated.

File: docs/designs/data_transform.md (hunk 13 above)
Comment (brightcoder01) [reply docs/designs/data_transform.md]: Good catch, updated.

File: docs/designs/data_transform.md (hunk 15 above)
Comment (brightcoder01) [reply docs/designs/data_transform.md]: Updated.

File: docs/designs/data_transform.md (hunk 14 above)
Comment (brightcoder01) [reply docs/designs/data_transform.md]: Updated.

File: docs/designs/data_transform.md (hunk 7 above)
Comment (brightcoder01) [reply docs/designs/data_transform.md]: In this doc, we focus on batch processing. Streaming data is not covered.

File: docs/designs/data_transform.md (hunk 8 above)
Comment (brightcoder01) [reply docs/designs/data_transform.md]: Thanks for the suggestion, I want to keep aligned with the statement of TFX Transform component. (https://www.tensorflow.org/tfx/guide/transform)

Comment (brightcoder01) [reply docs/designs/data_transform.md]: Good suggestion, updated.

File: docs/designs/data_transform.md (hunk 9 above)
Comment (brightcoder01) [reply docs/designs/data_transform.md]: Good suggestion, I've rephrase it.

**Review by terrytangyuan — APPROVED** [review]
